### PR TITLE
Stop leaking 'w' files

### PR DESCRIPTION
**Pull request by bboreham**
As noted in #1757, we left some copies of `weavewait`, copied as `w`, lying around.

This PR creates an auxilliary container `weavevolumes` to hold the `/w/w` files, so the proxy maps into that instead of creating new volumes every time.

Replaces #1768

### CODE REVIEW

**Comment by bboreham** [comment]
Cleaning up the `w` files breaks an assumption in test 195, that we can stop and restart the proxy and carry on relying on files mounted by the previous instance.

**Comment by awh** [comment]
> Cleaning up the w files breaks an assumption in test 195, that we can stop and restart the proxy and carry on relying on files mounted by the previous instance.

The assumption specifically is that the following sequence is valid:

1. `weave launch-proxy`
2. `docker run` container A via proxy
3. `docker stop` container A via proxy
4. `weave stop-proxy`
5. `weave start-proxy`
6. `docker start` container A via proxy

Which is completely reasonable, and worked before this PR, volume leak notwithstanding; the only reason the breakage wasn't noticed on first submission is because there was no test that exercised that sequence of steps until the `--no-multicast-route` PR got merged.

In other words, the failure is not due to a bad assumption in `195_no_multicast_route_1_test.sh` - it's because the leak fix has introduced a new problem.

**Comment by bboreham** [comment]
The "new" problem is that the system works if you leave files lying around forever.

I didn't say it was a bad assumption; however we need something creative to make progress.

**Comment by awh** [comment]
> The "new" problem is that the system works if you leave files lying around forever.

To my mind that's the existing problem, which this PR fixes at the cost of introducing a different one. Not sure which is worse though - slowly leaking volumes, or a hard failure in a probably-unfrequently-used sequence of steps.

> I didn't say it was a bad assumption

Yes, sorry, I inferred that erroneously :smile:

> we need something creative to make progress.

That may have to wait until tomorrow :smile:

**Comment by bboreham** [comment]
Maybe we can get round this by using a "volumes-only" container to hold on to `/w/w`, which sticks around until you do `weave reset`

(idea triggered by https://github.com/weaveworks/weave/issues/1859#issuecomment-172204574)

File: weave [hunk 1]
```diff
@@ -2102,6 +2105,7 @@ EOF
         for NAME in $CONTAINER_NAME $PROXY_CONTAINER_NAME ; do
             docker stop  $NAME >/dev/null 2>&1 || true
             docker rm -f $NAME >/dev/null 2>&1 || true
+            docker rm -v $VOLUMES_CONTAINER_NAME
```
Comment (rade) [inline weave]: this is rather wrong; you don't want this to be inside the loop, where it gets executed multiple times.

**Comment by bboreham** [comment]
Better now.

File: weave [hunk 2]
```diff
@@ -1670,8 +1671,13 @@ launch_proxy() {
     docker_client_args $DOCKER_CLIENT_ARGS
     proxy_args "$@"
     mkdir -p /var/run/weave
+    # Create a data-only container to mount the weavewait files from
+    if ! docker inspect -f ' ' $VOLUMES_CONTAINER_NAME > /dev/null 2>&1 ; then
+       docker create -v /w -v /w-noop -v /w-nomcast --name=$VOLUMES_CONTAINER_NAME --entrypoint=/w/w $EXEC_IMAGE >/dev/null
```
Comment (rade) [inline weave]: I find the `/w/w` obscure. `/bin/false`?

**Comment by rade** [comment]
This does fix #1757, right?

**Comment by bboreham** [comment]
An excellent question.

**Comment by bboreham** [comment]
In general it does, but our smoke-tests are masking the effect because they remove every container with `docker rm -f` before doing `weave reset`.

**Comment by bboreham** [comment]
I added `-v` to the `docker rm` in `config.sh`.  It's a bit of a special-case because we are going through contortions in `start_suite` to avoid hanging Docker.

In short, yes it fixes #1757.

**Comment by bboreham** [comment]
Maybe this should go into 1.4.3?

**Comment by rade** [comment]
> Maybe this should go into 1.4.3?

I was wondering about that too. The change is quite simple, though it does introduce a new container. I cannot think of any upgrade issues. So I suggest we put this in 1.4.3.

**Comment by bboreham** [comment]
Replaced by #1935 on 1.4 branch
